### PR TITLE
WFLY-6941 Split Weld subsystem

### PR DESCRIPTION
**Pull request by mkouba**
https://issues.jboss.org/browse/WFLY-6941

* split subsystem into core and optional modules (ejb, jpa, bean-validation, etc.)
* define SPI
* use ServiceLoader to load optional services

This a WildFly Swarm requirement - see also https://issues.jboss.org/browse/SWARM-568.

Both the Weld test suite and CDI TCK passed (verified manually).

### CODE REVIEW

File: feature-pack/src/main/resources/modules/system/layers/base/org/jboss/as/weld/transactions/main/module.xml [hunk 1]
```diff
@@ -0,0 +1,51 @@
+<?xml version="1.0" encoding="UTF-8"?>
+
+<!--
+  ~ JBoss, Home of Professional Open Source.
+  ~ Copyright 2010, Red Hat, Inc., and individual contributors
```
Comment (scottmarlow) [inline feature-pack/src/main/resources/modules/system/layers/base/org/jboss/as/weld/transactions/main/module.xml]: Copyright dates should be 2016

File: feature-pack/src/main/resources/modules/system/layers/base/org/jboss/as/weld/ejb/main/module.xml [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  ~ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  ~ Lesser General Public License for more details.
+  ~
+  ~ You should have received a copy of the GNU Lesser General Public
+  ~ License along with this software; if not, write to the Free
+  ~ Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+  ~ 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+  -->
+
+<module xmlns="urn:jboss:module:1.3" name="org.jboss.as.weld.ejb">
```
Comment (rhusar) [inline feature-pack/src/main/resources/modules/system/layers/base/org/jboss/as/weld/ejb/main/module.xml]: Should use latest modules version.

**Comment by scottmarlow** [comment]
@stuartwdouglas do you think that this be merged?  https://issues.jboss.org/browse/JBEAP-6913 doesn't appear to be acked yet, should we ask for it be?

**Comment by mkouba** [comment]
@scottmarlow We have pm_ack `+` on JBEAP-6913 now. I'm going to send a PR to EAP7 today and rebase and resolve conflicting files in this PR afterwards.

**Comment by mkouba** [comment]
Rebased.

File: weld/ejb/src/main/java/org/jboss/as/weld/ejb/StatefulSessionObjectReferenceImpl.java [hunk 3]
```diff
@@ -77,7 +79,8 @@ public StatefulSessionObjectReferenceImpl(EjbDescriptorImpl<?> descriptor) {
     @SuppressWarnings({ "unchecked" })
     public synchronized <S> S getBusinessObject(Class<S> businessInterfaceType) {
         if (isRemoved()) {
-            throw WeldLogger.ROOT_LOGGER.ejbHashBeenRemoved();
+            // Cannot be placed in WeldLogger
+            throw new NoSuchEJBException("EJB has been removed");
```
Comment (kabir) [inline weld/ejb/src/main/java/org/jboss/as/weld/ejb/StatefulSessionObjectReferenceImpl.java]: It needs i18n still.

You probably just need to create a new logger class in this weld/ejb module

**Comment by kabir** [comment]
This is a nitpick, but rather than putting WeldLogger into the logging package, and excluding that in module.xml, can it be put in a _private package similar to https://github.com/wildfly/wildfly/blob/master/feature-pack/src/main/resources/modules/system/layers/base/org/wildfly/extension/batch/jberet/main/module.xml#L31? It is slightly more standard I think. Do you agree @jamezp ?

**Comment by mkouba** [comment]
@kabir I'm sorry but I don't understand. Where exactly is `WeldLogger` excluded? Also I think it has been in logging package since the beginning.

File: weld/ejb/src/main/java/org/jboss/as/weld/ejb/StatefulSessionObjectReferenceImpl.java (hunk 3 above)
Comment (mkouba) [inline weld/ejb/src/main/java/org/jboss/as/weld/ejb/StatefulSessionObjectReferenceImpl.java]: Well, `javax.ejb.NoSuchEJBException` semantics is clearly defined by the EJB spec, that's why I thought it's not worth creating a new logger just for this exception. But I can do this if it's a requirement.

**Comment by kabir** [comment]
@mkouba It is excluded in https://github.com/wildfly/wildfly/pull/9324/files#diff-eb631e993f8c08924554ecd3baea9f81R32 which is fine. It would be a bit better if that package is called _private. I am not insisting on it, and only mentioned it since there is that other issue we are discussing.

File: weld/ejb/src/main/java/org/jboss/as/weld/ejb/StatefulSessionObjectReferenceImpl.java (hunk 3 above)
Comment (kabir) [inline weld/ejb/src/main/java/org/jboss/as/weld/ejb/StatefulSessionObjectReferenceImpl.java]: I think the logger should be added.

**Comment by mkouba** [comment]
@kabir I see. Well, `WeldLogger` is actually located in `wildfly-weld-common` module and shared by all modules now. So we should probably remove the aforementioned exclude from subsystem module.xml completely. Now I wonder whether it's a problem that `WeldLogger` is exported from `wildfly-weld-common`...

**Comment by jamezp** [comment]
Those test failures are related to the changes. It looks like there is a missing dependency.

**Comment by kabir** [comment]
@mkouba Add 
```
        <module name="org.jboss.logging"/>
```
As the first dependency of the ejb module.

For the record in response to my earlier comments about the shared logger, this is fine, it is in a private module.

**Comment by mkouba** [comment]
@kabir @jamezp Ah, yes of course. Dependency added...

**Comment by scottmarlow** [comment]
thanks this change! :)

**Pull request merged**